### PR TITLE
HBASE-23155 May NPE when concurrent AsyncNonMetaRegionLocator#updateC…

### PR DESCRIPTION
**Pull request by infraio**
…achedLocationOnError

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   2m  1s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  1s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 45s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   8m 16s |  master passed  |
| :green_heart: |  compile  |   1m 46s |  master passed  |
| :green_heart: |  checkstyle  |   2m 16s |  master passed  |
| :green_heart: |  shadedjars  |   6m  6s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   1m 11s |  master passed  |
| :blue_heart: |  spotbugs  |   5m 28s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   6m 44s |  master passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   7m  2s |  the patch passed  |
| :green_heart: |  compile  |   1m 42s |  the patch passed  |
| :green_heart: |  javac  |   1m 42s |  the patch passed  |
| :green_heart: |  checkstyle  |   2m 15s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 45s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  23m 22s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   1m 20s |  the patch passed  |
| :green_heart: |  findbugs  |   7m 13s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  |   2m  4s |  hbase-client in the patch passed.  |
| :broken_heart: |  unit  | 256m 53s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 48s |  The patch does not generate ASF License warnings.  |
|  |   | 347m 19s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestMasterShutdown |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-718/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/718 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux b5e574ab8ad9 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-718/out/precommit/personality/provided.sh |
| git revision | master / 5848e149f4 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-718/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-718/1/testReport/ |
| Max. process+thread count | 4495 (vs. ulimit of 10000) |
| modules | C: hbase-client hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-718/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestAsyncNonMetaRegionLocator.java [hunk 1]
```diff
@@ -399,4 +399,14 @@ public void testLocateBeforeInOnlyRegion() throws IOException, InterruptedExcept
...
     assertArrayEquals(loc.getRegion().getEndKey(), EMPTY_END_ROW);
   }
+
+  @Test
+  public void testConcurrentUpdateCachedLocationOnError() throws Exception {
+    createSingleRegionTable();
+    HRegionLocation loc =
+        getDefaultRegionLocation(TABLE_NAME, EMPTY_START_ROW, RegionLocateType.CURRENT, false)
+            .get();
+    IntStream.range(0, 100).parallel()
```
Comment (saintstack) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestAsyncNonMetaRegionLocator.java]: If single-region table, will this produce a null RegionLocation? Should there be something going on in the background clearing tableCache?

Comment (infraio) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestAsyncNonMetaRegionLocator.java]: The updateCachedLocationOnError will produce a null RegionLocation.

**Comment by guangxuCheng** [comment]
LGTM +1

**Pull request merged**